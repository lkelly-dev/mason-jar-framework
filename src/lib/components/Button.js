import React, { useContext } from 'react';
import { cx, css } from 'emotion';
import MODIFIERS from './Modifiers';
import { debugAttrs, debugMode } from '../utils/debugging';

const ButtonOverride = css`
  -webkit-appearance: inherit !important;
  font-weight: bold;
  box-shadow: none !important;
`;

const Button = props => {
  const { context, children, color, outline, size, status, shape, className, as, ...rest } = props;
  const c =  (context) ? useContext(context) : null;
  const debug = debugMode(c)
  const buttonColor = MODIFIERS.COLOR[color] ? `btn-${color}` : null;
  const outlineColor = MODIFIERS.COLOR[outline] ? `btn-outline-${outline}` : null;
  const buttonSize = MODIFIERS.SIZE[size] ? `btn-${size}` : null;
  const buttonShape = MODIFIERS.SHAPE[shape] ? `btn-${shape}` : null;
  const buttonStatus = MODIFIERS.BUTTON[status] ? `${status}` : null;
  const Component = as || 'button';
  const buttonClass = cx(
    'btn',
    buttonColor,
    outlineColor,
    buttonSize,
    buttonShape,
    buttonStatus,
    ButtonOverride,
    className,
  );
  return (
    <Component type="button" className={buttonClass}
      {...(debug) ? debugAttrs(props): null}
      {...rest}>
      {children}
    </Component>
  );
};

const ButtonGroup = props => {
  const { context, children, size, orientation, className, ...rest } = props;
  const c =  (context) ? useContext(context) : null;
  const debug = debugMode(c)
  const buttonGroupSize = MODIFIERS.SIZE[size] ? `btn-group-${size}` : null;
  const buttonGroupOrientation = MODIFIERS.ORIENTATION[orientation] ? `btn-group-${orientation}` : null;
  const buttonGroupClass = cx('btn-group', buttonGroupSize, buttonGroupOrientation, className);
  return (
    <div className={buttonGroupClass}
      {...(debug) ? debugAttrs(props): null}
      {...rest}>
      {children}
    </div>
  );
};

Button.Group = ButtonGroup;

export { Button, ButtonGroup };
