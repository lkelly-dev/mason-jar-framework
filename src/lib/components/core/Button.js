import React from 'react';
import { cx, css } from 'emotion';
import MODIFIERS from './Modifiers';

const ButtonOverride = css`
  -webkit-appearance: inherit !important;
  font-weight: bold;
  box-shadow: none !important;
`;

const Button = props => {
  const { children, color, outline, size, status, shape, className, as, ...rest } = props;
  const buttonColor = MODIFIERS.COLOR[color] ? `btn-${color}` : null;
  const outlineColor = MODIFIERS.COLOR[outline] ? `btn-outline-${outline}` : null;
  const buttonSize = MODIFIERS.SIZE[size] ? `btn-${size}` : null;
  const buttonShape = MODIFIERS.SHAPE[shape] ? `btn-${shape}` : null;
  const buttonStatus = MODIFIERS.BUTTON[status] ? `${status}` : null;
  const Component = as || 'button';
  const buttonClass = cx(
    'btn',
    buttonColor,
    outlineColor,
    buttonSize,
    buttonShape,
    buttonStatus,
    ButtonOverride,
    className,
  );
  return (
    <Component type="button" className={buttonClass} {...rest}>
      {children}
    </Component>
  );
};

const ButtonGroup = props => {
  const { children, size, orientation, className, ...rest } = props;
  const buttonGroupSize = MODIFIERS.SIZE[size] ? `btn-group-${size}` : null;
  const buttonGroupOrientation = MODIFIERS.SIZE[orientation] ? `btn-group-${orientation}` : null;
  const buttonGroupClass = cx('btn-group', buttonGroupSize, buttonGroupOrientation, className);
  return (
    <div className={buttonGroupClass} {...rest}>
      {children}
    </div>
  );
};

Button.Group = ButtonGroup;

export { Button, ButtonGroup };
