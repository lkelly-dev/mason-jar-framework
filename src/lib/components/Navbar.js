import React from 'react';
import { cx } from 'emotion';
import { Button } from './Button';
import MODIFIERS from './Modifiers';
import { debugAttrs, debugMode } from '../utils/debugging';

const Navbar = props => {
  const { context, children, expand, color, fixed, vertical, className, ...rest } = props;
  const debug = debugMode(context)
  const navbarColor = MODIFIERS.COLOR[color] ? `navbar-${color}` : null;
  const expandSize = MODIFIERS.SIZE[expand] ? `navbar-expand-${expand}` : null;
  const navbarFixed = MODIFIERS.FIXED[fixed] ? `fixed-${fixed}` : null;
  const navbarVertical = vertical ? 'navbar-vertical' : null;
  const navbarClass = cx('navbar', navbarColor, expandSize, navbarFixed, navbarVertical, className);
  return (
    <nav className={navbarClass}
      {...(debug && debugAttrs(props)) }
      {...rest}>
      {children}
    </nav>
  );
};

const NavbarNav = props => {
  const { context, children, className, ...rest } = props;
  const debug = debugMode(context)
  const navbarClass = cx('navbar-nav', className);
  return (
    <nav className={navbarClass}
      {...(debug && debugAttrs(props)) }
      {...rest}>
      {children}
    </nav>
  );
};

const NavbarBrand = props => {
  const { context, children, className, ...rest } = props;
  const debug = debugMode(context)
  const navBrandClass = cx('navbar-brand', className);
  return (
    <a className={navBrandClass}
      {...(debug && debugAttrs(props)) }
      {...rest}>
      {children}
    </a>
  );
};

const NavbarCollapse = props => {
  const { context, children, className, ...rest } = props;
  const debug = debugMode(context)
  const navbarCollapseClass = cx('collapse', 'navbar-collapse', className);
  return (
    <div className={navbarCollapseClass}
      {...(debug && debugAttrs(props)) }
      {...rest}>
      {children}
    </div>
  );
};

const NavbarToggle = props => {
  const { context, className, ...rest } = props;
  const togglerClass = cx('navbar-toggler', className);
  const debug = debugMode(context)
  return (
    <Button className={togglerClass}
      {...(debug && debugAttrs(props)) }
      {...rest}>
      <span className="navbar-toggler-icon" />
    </Button>
  );
};

const NavbarText = props => {
  const { context, children, className, ...rest } = props;
  const debug = debugMode(context)
  const navbarClass = cx('navbar-text', className);
  return (
    <span className={navbarClass}
      {...(debug && debugAttrs(props)) }
      {...rest}>
      {children}
    </span>
  );
};

Navbar.Nav = NavbarNav;
Navbar.Brand = NavbarBrand;
Navbar.Text = NavbarText;
Navbar.Collapse = NavbarCollapse;
Navbar.Toggle = NavbarToggle;

export { Navbar, NavbarNav, NavbarBrand, NavbarText, NavbarCollapse, NavbarToggle };
