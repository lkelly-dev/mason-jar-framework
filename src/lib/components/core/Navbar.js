import React from 'react';
import { cx } from 'emotion';
import { Button } from './Button';
import MODIFIERS from './Modifiers';

const Navbar = props => {
  const { children, expand, color, fixed, vertical, className, ...rest } = props;
  const navbarColor = MODIFIERS.COLOR[color] ? `navbar-${color}` : null;
  const expandSize = MODIFIERS.SIZE[expand] ? `navbar-expand-${expand}` : null;
  const navbarFixed = MODIFIERS.FIXED[fixed] ? `fixed-${fixed}` : null;
  const navbarVertical = MODIFIERS.ORIENTATION[vertical] ? 'navbar-vertical' : null;
  const navbarClass = cx('navbar', navbarColor, expandSize, navbarFixed, navbarVertical, className);
  return (
    <nav className={navbarClass} {...rest}>
      {children}
    </nav>
  );
};

const NavbarNav = props => {
  const { children, className, ...rest } = props;
  const navbarClass = cx('navbar-nav', className);
  return (
    <nav className={navbarClass} {...rest}>
      {children}
    </nav>
  );
};

const NavbarBrand = props => {
  const { children, className, ...rest } = props;
  const navBrandClass = cx('navbar-brand', className);
  return (
    <a className={navBrandClass} {...rest}>
      {children}
    </a>
  );
};

const NavbarCollapse = props => {
  const { children, className, ...rest } = props;
  const navbarCollapseClass = cx('collapse', 'navbar-collapse', className);
  return (
    <div className={navbarCollapseClass} {...rest}>
      {children}
    </div>
  );
};

const NavbarToggle = props => {
  const { className, ...rest } = props;
  const togglerClass = cx('navbar-toggler', className);
  return (
    <Button className={togglerClass} {...rest}>
      <span className="navbar-toggler-icon" />
    </Button>
  );
};

const NavbarText = props => {
  const { children, className, ...rest } = props;
  const navbarClass = cx('navbar-text', className);
  return (
    <span className={navbarClass} {...rest}>
      {children}
    </span>
  );
};

Navbar.Nav = NavbarNav;
Navbar.Brand = NavbarBrand;
Navbar.Text = NavbarText;
Navbar.Collapse = NavbarCollapse;
Navbar.Toggle = NavbarToggle;

export { Navbar, NavbarNav, NavbarBrand, NavbarText, NavbarCollapse, NavbarToggle };
