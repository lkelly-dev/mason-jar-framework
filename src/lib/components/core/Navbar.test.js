import React from 'react';
import { shallow } from 'enzyme';
import { Navbar, NavbarNav, NavbarBrand, NavbarText, NavbarCollapse, NavbarToggle } from "./Navbar";
import { Button } from './Button';

describe('Navbar', () => {
  it('Navbar matches snapshot', () => {
    const component = shallow(<Navbar></Navbar>);
    expect(component).toMatchSnapshot();
  })
  it('has expected basic props', () => {
    const cn = 'neato';
    const children = <div> children </div>;
    const component = shallow(
      <Navbar
        what="ever"
        className={ cn }>
        { children }
      </Navbar>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toHaveLength(2);
    expect(classNameArray).toContain('navbar');
    expect(classNameArray).toContain(cn);
    expect(component.prop('what')).toBe('ever');
    expect(component.type()).toBe('nav')
  })
  it('has expected complex props', () => {
    const cn = 'neato',
          expand = 'sm',
          color = 'warning',
          fixed = 'top',
          vertical = 'vertical';
    const children = <div> children </div>;
    const component = shallow(
      <Navbar
        what="ever"
        color={ color }
        expand={ expand }
        fixed={ fixed }
        vertical={ vertical }
        className={ cn }>
        { children }
      </Navbar>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('navbar');
    expect(classNameArray).toContain(`navbar-${color}`);
    expect(classNameArray).toContain(`navbar-expand-${expand}`);
    expect(classNameArray).toContain(`fixed-${fixed}`);
    expect(classNameArray).toContain('navbar-vertical');
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(6);
    expect(component.prop('what')).toBe('ever');
    expect(component.type()).toBe('nav')
  })
});



/*
describe('ModalClose', () => {
  it('matches snapshot', () => {
    const component = shallow(<ModalClose></ModalClose>);
    expect(component).toMatchSnapshot();
  })
  it('has expected props', () => {
    const cn = 'neato';
    const component = shallow(
      <ModalClose
        what="ever"
        className={ cn }>
      </ModalClose>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toHaveLength(2);
    expect(classNameArray).toContain('close');
    expect(classNameArray).toContain(cn);
    expect(component.prop('what')).toBe('ever');
    expect(component.prop('type')).toBe('button');
    expect(component.prop('data-dismiss')).toBe('modal');
    expect(component.prop('aria-label')).toBe('Close');
    expect(component.type()).toBe('button')
  })
});
*/

describe('NavbarNav', () => {
  it('Navbar matches snapshot', () => {
    const component = shallow(<NavbarNav></NavbarNav>);
    expect(component).toMatchSnapshot();
  })
  it('has expected props', () => {
    const cn = 'neato';
    const children = <div> children </div>;
    const component = shallow(
      <NavbarNav
        what="ever"
        className={ cn }>
        { children }
      </NavbarNav>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('navbar-nav');
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(2);
    expect(component.prop('what')).toBe('ever');
    expect(component.contains(children)).toEqual(true);
    expect(component.type()).toBe('nav');
  });
});

describe('NavbarBrand', () => {
  it('Navbar matches snapshot', () => {
    const component = shallow(<NavbarBrand></NavbarBrand>);
    expect(component).toMatchSnapshot();
  })
  it('has expected props', () => {
    const cn = 'neato';
    const children = <div> children </div>;
    const component = shallow(
      <NavbarBrand
        what="ever"
        className={ cn }>
        { children }
      </NavbarBrand>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('navbar-brand');
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(2);
    expect(component.prop('what')).toBe('ever');
    expect(component.contains(children)).toEqual(true);
    expect(component.type()).toBe('a');
  });
});

describe('NavbarCollapse', () => {
  it('Navbar matches snapshot', () => {
    const component = shallow(<NavbarCollapse></NavbarCollapse>);
    expect(component).toMatchSnapshot();
  })
  it('has expected props', () => {
    const cn = 'neato';
    const children = <div> children </div>;
    const component = shallow(
      <NavbarCollapse
        what="ever"
        className={ cn }>
        { children }
      </NavbarCollapse>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('collapse');
    expect(classNameArray).toContain('navbar-collapse');
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(3);
    expect(component.prop('what')).toBe('ever');
    expect(component.contains(children)).toEqual(true);
    expect(component.type()).toBe('div');
  });
});

describe('NavbarToggle', () => {
  it('Navbar matches snapshot', () => {
    const component = shallow(<NavbarToggle></NavbarToggle>);
    expect(component).toMatchSnapshot();
  })
  it('has expected props', () => {
    const cn = 'neato';
    const component = shallow(
      <NavbarToggle
        what="ever"
        className={ cn }>
      </NavbarToggle>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('navbar-toggler');
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(2);
    expect(component.prop('what')).toBe('ever');
    // expect(component.contains(children)).toEqual(true);
    expect(component.type()).toBe(Button);
  });
});

describe('NavbarText', () => {
  it('Navbar matches snapshot', () => {
    const component = shallow(<NavbarText></NavbarText>);
    expect(component).toMatchSnapshot();
  })
  it('has expected props', () => {
    const cn = 'neato';
    const children = <div>children</div>
    const component = shallow(
      <NavbarText
        what="ever"
        className={ cn }>
        { children }
      </NavbarText>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('navbar-text');
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(2);
    expect(component.prop('what')).toBe('ever');
    expect(component.contains(children)).toEqual(true);
    expect(component.type()).toBe('span');
  });
});
