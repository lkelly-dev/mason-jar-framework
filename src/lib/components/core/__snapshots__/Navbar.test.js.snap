// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Navbar matches snapshot 1`] = `
<nav
  className="navbar"
/>
`;

exports[`NavbarBrand Navbar matches snapshot 1`] = `
<a
  className="navbar-brand"
/>
`;

exports[`NavbarCollapse Navbar matches snapshot 1`] = `
<div
  className="collapse navbar-collapse"
/>
`;

exports[`NavbarNav Navbar matches snapshot 1`] = `
<nav
  className="navbar-nav"
/>
`;

exports[`NavbarText Navbar matches snapshot 1`] = `
<span
  className="navbar-text"
/>
`;

exports[`NavbarToggle Navbar matches snapshot 1`] = `
<Button
  className="navbar-toggler"
>
  <span
    className="navbar-toggler-icon"
  />
</Button>
`;
